test/Test/Browser.purs

```diff
@@ -82,17 +93,58 @@ runTests = do
     assert :: String -> Boolean -> Eff Effects Unit
     assert testName = assertEq testName true
 
-  _ <- hashes case _, _ of
+  k { assert, assertEq }
+
+
+runHashTests :: Eff Effects Unit -> Eff Effects Unit
+runHashTests next = withTest \{ assert } -> do
+  doneRef <- newRef (pure unit)
+  let done = join (readRef doneRef) *> next
+  writeRef doneRef =<< hashes case _, _ of
     Nothing, ""   -> assert "Hashes: Initial value" true
     Just "", "a"  -> assert "Hashes: ? -> a" true *> setHash "b"
     Just "a", "b" -> assert "Hashes: a -> b" true *> setHash ""
-    Just "b", ""  -> assert "Hashes: b -> ?" true
+    Just "b", ""  -> assert "Hashes: b -> ?" true *> done
     _, _          -> assert "Hashes: fail" false
-
   setHash "a"
 
+runPushStateTests :: Eff Effects Unit
+runPushStateTests = withTest \{ assert } -> do
+  hist <- makeInterface (Basename "/test")
+  doneRef <- newRef (pure unit)
+  let
+    done = join (readRef doneRef)
+    readState r = r { state = hush $ runExcept $ readInt r.state }
+    loc1 = { state: Nothing, pathname: "/", search: "", hash: "", path: "/" }
+    loc2 = { state: Just 1, pathname: "/a", search: "?a", hash: "", path: "/a?a" }
+    loc3 = { state: Just 2, pathname: "/b", search: "", hash: "#b", path: "/b#b" }
+    loc4 = { state: Just 3, pathname: "", search: "", hash: "", path: "" }
+
+  writeRef doneRef =<< flip locations hist \old new ->
+    case readState <$> old, readState new of
+      Nothing, new'
+        | Rec.equal new' loc1 -> do
+          assert "Locations: Initial" true
+      Just old', new'
+        | Rec.equal old' loc1 && Rec.equal new' loc2 -> do
+          assert "Locations: init -> b" true
+          hist.pushState (toForeign 2) (Path "/b#b")
+        | Rec.equal old' loc2 && Rec.equal new' loc3 -> do
```

Comment:
So, if I want to make a data type which has a tag for each route: `data MyRoute = Home | MyProfile | ItemPage ItemName | FriendProfile UserName | NotFound`, I'd need to write a function `LocationState -> MyRoute`? I ask because that's the interface I see people recommend.

Reply:
You would use `matches` for that, just like with the hash-based interface, which takes a `Match` parser. If you need to parse routes _and_ use the state then you would need to manually run your parser. I can possibly add `matchesWithState` to yield the `Foreign` state.